src/redprl/refiner_types.fun

```diff
@@ -150,68 +171,93 @@ struct
     fun EqFCom alpha jdg =
       let
         val _ = RedPrlLog.trace "WBool.EqFCom"
-        val (I, H) >> CJ.EQ ((lhs, rhs), ty) = jdg
+        val (I, H) >> CJ.EQ ((lhs, rhs), (ty, k)) = jdg
         val Syn.WBOOL = Syn.out ty
+        val _ = Assert.kindLeq (inherentKind, k)
         val Syn.FCOM args0 = Syn.out lhs
         val Syn.FCOM args1 = Syn.out rhs
       in
-        ComKit.EqFComDelegator alpha (I, H) args0 args1 ty
+        ComKit.EqFComDelegator alpha (I, H) args0 args1 (ty, k)
       end
 
     fun Elim z _ jdg =
       let
         val _ = RedPrlLog.trace "WBool.Elim"
-        val (I, H) >> CJ.TRUE cz = jdg
-        val CJ.TRUE ty = lookupHyp H z
+        val (I, H) >> CJ.TRUE (cz, k) = jdg
+        (* if(FCOM) steps to COM *)
+        val k = K.meet (k, K.COM)
+        (* for now we ignore the kind in the context *)
+        val CJ.TRUE (ty, _) = Hyps.lookup z H
         val Syn.WBOOL = Syn.out ty
+        
+        (* kind because of FCOM *)
+        val goalKind = makeTypeIfLess (I, H) (cz, k) K.top
```

Comment:
If the failure is due to this rule, the goal `goalKind` is actually unconditional. So I guess the tactic assumes this rule only generates two goals? Now this rule will generates three, always.